### ISSUE TITLE
Implement metrics for swarm

### ISSUE THREAD

**Issue by holisticode** [ethersphere/swarm#183]
This issue has been opened in order to be tracked in the ethersphere project. It's source issue is https://github.com/ethereum/go-ethereum/issues/15481:

> For swarm, it would be good to be able to collect stats and metrics on a node concerning storage (chunks, local DB, etc.) and chequebook properties (consumed and delivered services, service peers, balances, cheques, etc.).

This issue is a parent issue for https://github.com/ethersphere/go-ethereum/issues/177, which is about evaluating the technical infrastructure for the metrics implementation.

Also related are issues https://github.com/ethersphere/go-ethereum/issues/159 and https://github.com/ethersphere/go-ethereum/issues/158 which may be visualizing information based on metrics

**Comment by holisticode** [comment]
As I see now, this task has different dimensions:

1. The infrastructure part. Which library to use? Which metrics format? What stack for catching - storing - visualizing metrics do we use? What model (e.g. push - pull)? What infrastructure (databases, scripts, servers, websites) need to be set up? A separate issue has already been created to track this: https://github.com/ethersphere/go-ethereum/issues/177
1. The code part. This is about WHICH metrics will be collected and HOW to collect and structure them. This issue will focus on this.

**Comment by holisticode** [comment]
Also for the code part, we could distinguish different dimensions:

  1. Development metrics. These metrics will be collected in order to get insights on the code and improve performance and function of swarm. They target the **the developer** and will be collected for **development** and **test**.
  1. User metrics. Such metrics are more useful for the **user**. Some of these metrics could be even classified as *private*. May include storage usage or all swap accounting related information. 
  1. Runtime metrics. More targeting to get statistical information about nodes, getting an overall sense of the swarm network, its state and dynamics. Could encompass start/stop events of nodes,number of current nodes in the network, cache deletion events, etc.

As usual, multi-dimensional challenges are rarely well-addressed by one-dimensional approaches like categorization. Concretely, many developer metrics may well also be runtime metrics. Thus, one idea on how to structure metrics is the usage of tags.

**Comment by holisticode** [comment]
Development metrics should be available always during development, enabled by a flag when running tests. The developer should be able to easily set up the infrastructure (through scripts) in order to collect and monitor the targeted metrics. There may be a predefined set of development metrics, but it should be easy for the developer to add metrics to tests and also to main code.

User metrics should be available by querying the running node. This could be via browser on a HTTP endpoint. The `admin` endpoint planned with issue https://github.com/ethersphere/go-ethereum/issues/159 is the best candidate for this.

Runtime metrics may well be collected and shown via the local node, but some if not all may also be interesting in being sent to a central location for (anonymous) visualization, like ethereum's `ethstats` page.

For the sake of simplicity, development and runtime metrics could be colluded into one single configuration, provided the developer and runtime-stakeholder needs are addressed by the chosen setup.

**Comment by holisticode** [comment]
What follows is a list of identified metrics. This is not a conclusive list and can be edited.

Development:
  * Number of peers -   `network.peers.num `(hive)
  * Kad table peer changes (addresses, number of peers in each bin) `network.kademlia.on`, `network.kademlia.bucket.onindex`, `network.kademlia.off`, `network.kademlia.bucket.offindex`
  * `Healthy` status counters (how many times)
  * Healthy timers (how long until healthy)
  * Number of peers dropeed  - `network-removepeer.count`
  * Number of peers added    - `network.addpeer.count`
  * DB size     - `storage.db.chunks.size`
  * Memstore (cache) size   - `storage.db.cache.size`
  * Syncing peers
  * Syncing data amount
  * Garbage collected addresses - `storage.db.dbstore.gc.address`
  * Number of garbage collected chunks - `storage.db.dbstore.gc.count`
  * Number of chunks created  - `storage.chunks.new`
  * Number of chunks stored    -  `storage.db.memstore.put.count`,   `storage.db.dbstore.put.count`
  * Number of chunks synced (sent/received) - `network.sync.recv.count`, `network.sync.send.count`
  * Number of chunks received but ignored    - `network.sync.recv.ignore`, `network.sync.send.refused`, `network.sync.send.notfound`
  * Number of chunks purged   - `storage.db.memstore.rm.count`, `storage.db.dbstore.delete.count`
  * Chunk addresses created  - `storage.chunks.addresses.new`
  * Chunk addresses stored    - `storage.db.memstore.put.address`, `storage.db.dbstore.put.address`
  * Chunk addresses purged   - `storage.db.memstore.rm.address`, `storage.db.dbstore.delete.address`
  * Chunk addresses received (syncing)
  * Number of messages        - `network.protocol.msg.storerequest.count`, `network.protocol.msg.retrieverequest.count`, `network.protocol.msg.peers.count`, `network.protocol.msg.syncrequest.count`, `network.protocol.msg.unsyncedkeys.count`, `network.protocol.msg.deliverrequest.count`, `network.protocol.msg.payment.count`, `network.protocol.msg.invalid.count`,  `network.protocol.msg.handlestatus.count`
  * Number of forwards
  * Forwarding addresses
  

User
  * Swap balances (global / per peer)
  * Swap payments received (amount/counter)
  * Swap payments sent (amount/counter)
  * Cheques counters (redeem/issue)
 


Runtime
   * Response times     - `http.request.response_time`
  * Number of uploads - `api.http.post.raw.count`, `api.http.post.raw.fail`, `api.http.post.files.count`, `api.http.post.files.fail`, `api.http.delete.count`, `api.http.delete.fail`
  * Size of uploads   
  * Number of retrievals -    `api.request.count` `api.get.count`, `api.get.notfound`, `api.http.get.list.count`, `api.http.get.list.fail`,`api.http.get.file.count`, `api.http.get.file.fail`,`api.http.get.file.notfound`,`api.http.get.file.fail`
  * Size of retrievals            -
  * Number of "edits" (PUT, DELETE)  
              - `api.put.count`, `api.put.fail`, `api.delete.count`,`api.delete.fail`, `api.modify,count`,`api.modify.fail`, `api.addfile.count`, `api.addfile.fail`, `api.removfile.count`,`api.removefile.fail`,  `api.appendfile.count`,`api.appendfile.fail`, `api.builddirtree.count`, `api.builddirtree.fail`
  * Number of successful resolves     - `api.resolve.count`
  * Number of failed resolves            -  `api.resolve.fail`
  * Number of Multiple Choices         -`api.get.http300`
  * Number of error pages served    - `api.http.errorpage.html.count`, `api.http.errorpage.json.count`
  * Start event count                         - `stack.start`
  * Node uptime  (currently only after stop) - `stack.uptime`
  * Stop event count                         - `stack.stop`
  * Counter for protocol requests (bzz, bzzh, bzzi)

**Comment by holisticode** [comment]
Probably such a differentiation of metrics can be implemented by assigning metrics to different `Registry` instances.

**Comment by nonsense** [comment]
@holisticode now that the metrics and instrumentation PRs are merged to master, I suggest we deploy the latest master to the cluster, and create a few dashboards with the current set of measurements.

We should then review what other measurements we want to add, and iterate further on this.

**Comment by holisticode** [comment]
I agree - however, I wonder if we shouldn't actually setup a test cluster with the new master first and see how it behaves there?

**Comment by gbalint** [comment]
@nonsense @holisticode Can this issue be closed? If we need new metrics imho we should create new issues for them

**Comment by nonsense** [comment]
Yes, we should review our dashboards and measurements and iteratively refine from here on.
